packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts

```diff
@@ -0,0 +1,92 @@
+import { GraphApi } from "@hashintel/hash-graph-client";
+import { ApolloError, UserInputError } from "apollo-server-express";
+import { types } from "@hashintel/hash-shared/types";
+import { VersionedUri } from "@blockprotocol/type-system-web";
+import { WORKSPACE_TYPES } from "../../../../graph/workspace-types";
+import { AccountFields, EntityModel, UserModel } from "../../../../model";
+
+const validateAccountShortname = async (
+  graphApi: GraphApi,
+  shortname: string,
+) => {
+  if (AccountFields.shortnameContainsInvalidCharacter(shortname)) {
+    throw new UserInputError(
+      "Shortname may only contain letters, numbers, - or _",
+    );
+  }
+  if (shortname[0] === "-") {
+    throw new UserInputError("Shortname cannot start with '-'");
+  }
+
+  if (
+    AccountFields.shortnameIsRestricted(shortname) ||
+    (await AccountFields.shortnameIsTaken(graphApi, { shortname }))
+  ) {
+    throw new ApolloError(`Shortname "${shortname}" taken`, "NAME_TAKEN");
+  }
+
+  /**
+   * @todo: enable admins to have a shortname under 4 characters
+   * @see https://app.asana.com/0/1201095311341924/1203285346775714/f
+   */
+  if (shortname.length < AccountFields.shortnameMinimumLength) {
+    throw new UserInputError("Shortname must be at least 4 characters long.");
+  }
+  if (shortname.length > AccountFields.shortnameMaximumLength) {
+    throw new UserInputError("Shortname cannot be longer than 24 characters");
+  }
+};
+
+type BeforeUpdateEntityHookCallback = (params: {
+  graphApi: GraphApi;
+  entityModel: EntityModel;
+  /**
+   * @todo: use improved typing of the properties object
+   * @see https://app.asana.com/0/1202805690238892/1203285029221334/f
+   */
+  updatedProperties: any;
+}) => Promise<void>;
```

Comment:
Sorry if I'm missing something from the other reviews, but shouldn't this be fallible and allow cancellation of the rest of the entity saving chain? Or is the kratos communication handled somewhere else (for example)?

Reply:
> shouldn't this be fallible and allow cancellation of the rest of the entity saving chain?

Yes, the way this would currently work is that an error would be thrown in the callback method which would ensure the properties aren't updated. Is there an alternative method for accomplishing this you had in mind?

Reply:
I was thinking that the method could be explicit about failing by its return value instead. But that isn't really much different from just throwing/catching, especially if we just throw it all the way to the Apollo handler.

Being explicit may also allow us to get rid of the dependency on `"apollo-server-express"`. But we can accomplish that by defining our own exceptions. Not sure any of these changes would bring much value, though. 

Perhaps it could allow us to retry rather than bail in the `updateProperties` method if we have explicit failures? But unsure we need that right now.

Reply:
> Being explicit may also allow us to get rid of the dependency on "apollo-server-express". But we can accomplish that by defining our own exceptions. Not sure any of these changes would bring much value, though.

These methods are specifically for GraphQL resolvers, so I'm not too concerned by this dependency. If we want to move away from apollo errors, wouldn't we would lose some of the GQL error handling functionality and I'm not sure what we would be accomplishing... but there's lots of room for improvement regarding how we currently throw errors, so definitely expect things to change.